src/cpp/session/SessionConsoleProcessSocket.cpp

```diff
@@ -0,0 +1,336 @@
+/*
+ * SessionConsoleProcessSocket.cpp
+ *
+ * Copyright (C) 2009-17 by RStudio, Inc.
+ *
+ * Unless you have received this program directly from RStudio pursuant
+ * to the terms of a commercial license agreement with RStudio, then
+ * this program is licensed to you under the terms of version 3 of the
+ * GNU Affero General Public License. This program is distributed WITHOUT
+ * ANY EXPRESS OR IMPLIED WARRANTY, INCLUDING THOSE OF NON-INFRINGEMENT,
+ * MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE. Please refer to the
+ * AGPL (http://www.gnu.org/licenses/agpl-3.0.txt) for more details.
+ *
+ */
+
+#include <session/SessionConsoleProcessSocket.hpp>
+
+#include <boost/make_shared.hpp>
+
+#include <core/FilePath.hpp>
+#include <core/json/Json.hpp>
+
+
+namespace rstudio {
+namespace session {
+namespace console_process {
+
+using namespace rstudio::core;
+
+namespace {
+
+// rapid reseeding via srand(time) causes same "random" sequence to be
+// returned by rand; only an issue for unit tests, really
+bool s_didSeedRand = false;
+
+} // anonymous namespace
+
+ConsoleProcessSocket::ConsoleProcessSocket()
+   :
+     port_(0),
+     serverRunning_(false),
+     activeConnections_(0)
+{
+}
+
+ConsoleProcessSocket::~ConsoleProcessSocket()
+{
+   try
+   {
+      stopServer();
+   }
+   catch (...)
+   {
+   }
+}
+
+Error ConsoleProcessSocket::ensureServerRunning()
+{
+   if (serverRunning_)
+      return Success();
+
+   long port = 0;
+   unsigned portRetries = 0;
+
+   // initialize seed for random port selection
+   if (!s_didSeedRand)
+   {
+      srand(time(NULL));
+      s_didSeedRand = true;
+   }
+
+   // no user-specified port; pick a random port
+   port = 3000 + (rand() % 5000);
+
+   try
+   {
+      pwsServer_.reset(new terminalServer());
+
+      pwsServer_->set_access_channels(websocketpp::log::alevel::none);
+      pwsServer_->init_asio();
+
+      pwsServer_->set_message_handler(
+               boost::bind(&ConsoleProcessSocket::onMessage, this, &*pwsServer_, _1, _2));
+      pwsServer_->set_http_handler(
+               boost::bind(&ConsoleProcessSocket::onHttp, this, &*pwsServer_, _1));
+      pwsServer_->set_close_handler(
+               boost::bind(&ConsoleProcessSocket::onClose, this, &*pwsServer_, _1));
+      pwsServer_->set_open_handler(
+               boost::bind(&ConsoleProcessSocket::onOpen, this, &*pwsServer_, _1));
+
+      // try to bind to the given port
+      do
+      {
+         try
+         {
+            if (core::FilePath("/proc/net/if_inet6").exists())
+            {
+               // listen will fail without ipv6 support on the machine so we
+               // only use it for machines with a ipv6 stack
+               // TODO (gary) need a cross-platform way to do this? does it
+               // matter on Mac/Windows clients?
+               pwsServer_->listen(port);
```

Comment:
When this fails because of a lack of ipv6 support, is it clear that lack of ipv6 support is the cause of the failure? If so seems like we could just always try ipv6 as a matter of course without attempting to sniff for it. (Alternately I'm a fan of just #ifdef for now)